### PR TITLE
Implement magazine retrieval

### PR DESCRIPTION
**Pull request by jjennifergu**
## Overview

Work in progress of building the Magazine model and creating appropriate functions + queries to display magazines on the Magazine page, Publication page, and Bookmarks page.

## Changes Made

- Built Magazine model in `Models/Magazine.swift`
- Created `Networking/MagazineQueries.graphql` to generate magazine queries
- Created fetch methods to retrieve magazines for the _Featured_ and _More Magazines_ sections of `MainView/MagazinesList.swift`
- Added configuration for `MagazineCell`

## Test Coverage

Tested queries on [dev endpoint](https://volume-dev.cornellappdev.com/graphql).

_Note: There is currently no magazine data on the backend. Will playtest after database is populated._

## Next Steps

- Support filtering by semester in _More Magazines_ section of Magazine Page
- Implement function to get magazine by publication for display on Publication Page
- Build PDF viewer and implement bookmarking for magazines

## Screenshots

No data yet, [Figma page](https://www.figma.com/file/fXgvLfYMD7Dgzw4QsEysJ2/Volume-FA22?node-id=58%3A1616) for reference (Pages -> Magazines [Final]).

<details>

  <summary>Coming Soon</summary>

  
  

</details>

### CODE REVIEW

File: Volume/Views/MainView/MagazineDetail.swift [hunk 1]
```diff
@@ -9,13 +9,14 @@
...
-struct MagazineDetail_Previews: PreviewProvider {
-    static var previews: some View {
-        MagazineDetail()
-    }
-}
+//struct MagazineDetail_Previews: PreviewProvider {
+//    static var previews: some View {
+//        MagazineDetail()
+//    }
+//}
```
Comment (ReadeP) [inline Volume/Views/MainView/MagazineDetail.swift:22]: Can you delete this commented code

File: Volume/Views/MainView/MagazinesList.swift [hunk 2]
```diff
@@ -10,23 +10,50 @@ import Combine
...
+        
+        fetchFeaturedMagazines(done)
+        fetchOtherMagazines()
+    }
+    
+    private func fetchFeaturedMagazines(_ done: @escaping () -> Void = { }) {
+        sectionQueries.featuredMagazines = Network.shared.publisher(for: GetFeaturedMagazinesQuery(limit: 7))
+            .map {
+                // not sure if force-unwrapping here is safe
+                $0.magazines!.map(\.fragments.magazineFields)
```
Comment (ReadeP) [inline Volume/Views/MainView/MagazinesList.swift:31]: Delete this comment too. I would try to not force unwrap if it you can. @dishanest thoughts?

File: Volume/Views/MainView/MagazinesList.swift [hunk 3]
```diff
@@ -10,23 +10,50 @@ import Combine
...
+            }
             .sink { completion in
                 networkState.handleCompletion(screen: .magazinesList, completion)
-            } receiveValue: {
+            } receiveValue: { magazineFields in
+                let featuredMagazines = [Magazine](magazineFields)
+
```
Comment (ReadeP) [inline Volume/Views/MainView/MagazinesList.swift:37]: delete newline

File: Volume/Views/MainView/MagazinesList.swift [hunk 4]
```diff
@@ -10,23 +10,50 @@ import Combine
...
+                let featuredMagazines = [Magazine](magazineFields)
+                
+                withAnimation(.linear(duration: 0.1)) {
+                    sectionStates.featuredMagazines = .results(featuredMagazines)
+                }
                 done()
+            }
+    }
+    
+    private var currentSemester = "fa22"
```
Comment (ReadeP) [inline Volume/Views/MainView/MagazinesList.swift:45]: Can this be defined inside a function and not a class value? If it has to be a class value though it should be defined at the top of the class

File: Volume/Views/MainView/MagazinesList.swift [hunk 5]
```diff
@@ -118,20 +145,24 @@ struct MagazinesList: View {
...
+        otherMagazines: AnyCancellable?
+    )
 }
 
-struct MagazinesList_Previews: PreviewProvider {
-    static var previews: some View {
-        MagazinesList()
-    }
-}
+//struct MagazinesList_Previews: PreviewProvider {
```
Comment (ReadeP) [inline Volume/Views/MainView/MagazinesList.swift:164]: Delete this commented out code

**Review by ReadeP — APPROVED** [review]
Great job! Looks really solid - just left a little feedback you should address before you merge

File: Volume/Views/MainView/MagazinesList.swift (hunk 2 above)
Comment (dishanest) [reply Volume/Views/MainView/MagazinesList.swift:31]: Chain the optionals

File: Volume/Views/MainView/MagazinesList.swift (hunk 4 above)
Comment (dishanest) [reply Volume/Views/MainView/MagazinesList.swift:45]: Probably better to write this in a `Constants` struct, e.g. 
```swift
private struct Constants {
   let currentSemester = "fa22"
}
```
Could also include other hard-coded values like the `0.1` animation duration, etc.

File: Volume/Views/MainView/MagazinesList.swift [hunk 6]
```diff
@@ -57,22 +84,22 @@ struct MagazinesList: View {
         }
     }
 
-    private var moreMagazinesSection: some View {
+    private var otherMagazinesSection: some View {
```
Comment (dishanest) [inline Volume/Views/MainView/MagazinesList.swift:87]: Could be more descriptive w/ the name, e.g. `magazinesBySemesterSection`?

File: Volume/Views/MainView/MagazineDetail.swift [hunk 7]
```diff
@@ -9,13 +9,14 @@
 import SwiftUI
 
 struct MagazineDetail: View {
```
Comment (dishanest) [inline Volume/Views/MainView/MagazineDetail.swift:11]: I would name this `MagazineReaderView` or something

File: Volume/Views/MainView/MagazinesList.swift [hunk 8]
```diff
@@ -57,22 +84,22 @@ struct MagazinesList: View {
...
                     case .loading:
                         ForEach(0..<10) { _ in
                              MagazineCell.Skeleton()
                         }
                     case .reloading(let results), .results(let results):
                         // TODO: Replace with results.trendingMagazines when backend is setup
-                        ForEach(0..<10) { _ in
-                            NavigationLink(destination: MagazineDetail()) {
+                        ForEach(results) { magazine in
+                            NavigationLink(destination: MagazineDetail(title: magazine.title)) {
```
Comment (dishanest) [inline Volume/Views/MainView/MagazinesList.swift:102]: Eventually you probably want to pass in the whole magazine object

File: Volume/Views/MainView/MagazinesList.swift [hunk 9]
```diff
@@ -118,20 +145,24 @@ struct MagazinesList: View {
 }
 
 extension MagazinesList {
-    typealias Results = (
-        trendingMagazines: [Magazine],
-        otherMagazines: [String : [Magazine]]
+    typealias ResultsPublisher = Publishers.Zip<
```
Comment (dishanest) [inline Volume/Views/MainView/MagazinesList.swift:148]: You're a legend

**Review by dishanest — CHANGES_REQUESTED** [review]
Sick first PR! Great adhesion to conventions & style. Just a few comments aside from @ReadeP's

File: Volume/Views/MainView/MagazineDetail.swift [hunk 10]
```diff
@@ -8,14 +8,9 @@
 
 import SwiftUI
 
-struct MagazineDetail: View {
+struct MagazineReaderView: View {
```
Comment (dishanest) [inline Volume/Views/MainView/MagazineDetail.swift:11]: make sure to change the file name too!

**Review by dishanest — APPROVED** [review]

**Pull request merged**